api/desecapi/serializers.py

```diff
@@ -218,12 +129,152 @@ def validate_type(value):
 
     def to_representation(self, instance):
         data = super().to_representation(instance)
-        data.pop('id')
+        data.pop('id')  # TODO remove this by changing Meta subclass
         return data
 
+    def create(self, validated_data):
+        rrs_data = validated_data.pop('records')
+        rrset = RRset.objects.create(**validated_data)
+        rrset.set_all_record_contents(set([rr_data['content'] for rr_data in rrs_data]))
+        return rrset
+
+    def update(self, instance: RRset, validated_data):
+        # TODO this is duplicate code from DRF -- what is this for?
+        #  probably related to bulk serializers!
+        raise_errors_on_nested_writes('update', self, validated_data)
+
+        rrs_data = validated_data.pop('records', None)
+        if rrs_data is not None:
+            instance.set_all_record_contents(set([rr_data['content'] for rr_data in rrs_data]))
+
+        ttl = validated_data.pop('ttl', None)
+        if ttl and instance.ttl != ttl:
+            instance.ttl = ttl
+            instance.save(update_fields={'ttl'})  # TODO is passing ttl needed here?
+
+        return instance
+
+
+class RRsetListSerializer(ListSerializer):
+
+    def to_internal_value(self, data):
+        # TODO this is mostly copy-and-paste from rest_framework.serializers.ListSerializer.to_internal_value!
+
+        # TODO needed?
+        # if html.is_html_input(data):
+        #     data = html.parse_html_list(data, default=[])
+
+        if not isinstance(data, list):
+            message = self.error_messages['not_a_list'].format(
+                input_type=type(data).__name__
+            )
+            raise ValidationError({
+                api_settings.NON_FIELD_ERRORS_KEY: [message]
+            }, code='not_a_list')
+
+        if not self.allow_empty and len(data) == 0:
+            if self.parent and self.partial:
+                raise SkipField()
+
+            message = self.error_messages['empty']
+            raise ValidationError({
+                api_settings.NON_FIELD_ERRORS_KEY: [message]
+            }, code='empty')
+
+        ret = []
+        errors = []
+
+        partial = self.partial
+
+        known_rrsets = {(x.subname, x.type) for x in self.instance} if isinstance(self.instance, list) else set()
+
+        for item in data:
+            try:
+                self.partial = partial and (item.get('subname', ''), item.get('type', None)) in known_rrsets
```

Comment:
interesting algorithm, with `self.partial` stored away and recovered later. Looks convincing to me though.